### PR TITLE
Using flash memory to store user-defined types

### PR DESCRIPTION
**Pull request by MartBent**
Hello,

I wrote a function that accepts a generic type parameter that enables the developer to flash its own types into the flash memory. The provided test case passed on my STM32 NUCLEO-WL55JC2 board. If the type is not aligned to 64 bits (after struct alignment) the function will add a padding with value 0x00 to the programmed double word.

I am new to (embedded) Rust as well as open source contributing so don't hesitate to provide me with feedback / suggestions.

Regards, 

Mart

### CODE REVIEW

**Comment by newAM** [comment]
Thanks for the pull request!  I will try to review it tonight if I have time, if not I will get to it this weekend.

**Review by newAM — COMMENTED** [review]
Thanks for adding a test with your changes :+1: !

I wanted to try this out myself so I pushed the request changes to your branch; feel free to revert/modify these.

I also added a test for zero-sized types, and added some handling in the function for this.

Let me know what you think of changes!

File: hal/src/flash.rs [hunk 1]
```diff
@@ -403,6 +406,81 @@ impl<'a> Flash<'a> {
         ret
     }
 
+    /// Program a user-defined type.
+    ///
+    /// # Safety
+    ///
+    /// 1. Do not write to flash memory that is being used for your code.
+    /// 2. The destination address must be within the flash memory region.
+    /// 3. The `from` and `to` pointers must be aligned to the pointee type.
```
[the remark this answers is not in the document]
Comment (MartBent) [reply hal/src/flash.rs]: I took the safety command of the standard_program function, since I didn't really know what the safety concerns would be for this function. This make a more sense.

File: hal/src/flash.rs [hunk 2]
```diff
@@ -403,6 +406,81 @@ impl<'a> Flash<'a> {
...
+        }
+
+        self.clear_all_err();
+
+        c1_c2!(
+            self.flash.cr.modify(|_, w| w.pg().set_bit()),
+            self.flash.c2cr.modify(|_, w| w.pg().set_bit())
+        );
+
+        unsafe {
```
[the remark this answers is not in the document]
Comment (MartBent) [reply hal/src/flash.rs]: I forgot to properly set the unsafe scope, this is definitely better.

File: hal/src/flash.rs [hunk 3]
```diff
@@ -403,6 +406,81 @@ impl<'a> Flash<'a> {
...
+            //Get the size of the type T in bytes.
+            let size = core::mem::size_of::<T>() as isize;
+
+            //Amount of bytes that are already written.
+            let mut written_bytes = 0;
+
+            //Calculate the amount of double words needed to program the whole type.
+            let double_words = (size as f32 / 8.0).ceil() as isize;
```
[the remark this answers is not in the document]
Comment (MartBent) [reply hal/src/flash.rs]: Good change.

File: hal/src/flash.rs [hunk 4]
```diff
@@ -403,6 +406,81 @@ impl<'a> Flash<'a> {
...
+
+            //Write the type as double words
+            for n in 0..double_words - 1 {
+                write_volatile(
+                    (to as *mut u64).offset(n),
+                    (from as *const u64).offset(n).read(),
+                );
+                written_bytes += 8;
+            }
```
[the remark this answers is not in the document]
Comment (MartBent) [reply hal/src/flash.rs]: I didn't know Rust had this fold function, good change, this makes the code a lot cleaner.

File: hal/src/flash.rs [hunk 5]
```diff
@@ -403,6 +406,81 @@ impl<'a> Flash<'a> {
...
+            //Determine how many bytes are left to write
+            let bytes_left = size - written_bytes;
+
+            let mut last_double_word = 0;
+
+            //Append the left over bytes to a double word, the last few bytes can look random in flash memory since Rust uses memory alignment to make accessing faster.
+            for n in 0..bytes_left {
+                let byte = (from as *const u8).offset(written_bytes + n).read();
+                last_double_word |= (byte as u64) << (8 * n);
+            }
```
[the remark this answers is not in the document]
Comment (MartBent) [reply hal/src/flash.rs]: Again, a cleaner code

**Comment by MartBent** [comment]
You made good changes. I knew the code could use some improvement, thanks for the feedback en explanations!
